### ISSUE TITLE
[Improvement] Implement FidelityFX Super Resolution as an improvement over the bilinear(?) upscaling that Ren'Py uses

### ISSUE THREAD

**Issue by NintendoManiac64** [renpy/renpy#2909]
AMD's FidelityFX Super Resolution was released as open source today:
https://github.com/GPUOpen-Effects/FidelityFX-FSR

And the only thing on my mind is how it could work as an improvement over the bilinear(?) upscaling that Ren'Py currently uses for when then program window is higher resolution than a given art asset/video without being a complete performance hog on lower-end integrated GPU solutions (e.g. Westmere, Sandy Bridge, etc)

...though for specifically anime-styled art, I imagine that the "[Anime4K](https://github.com/bloc97/Anime4K)" and "[Anime4KCCP](https://github.com/TianZerL/Anime4KCPP)" would probably give better visual results, but I would expect it to be more computationally-heavy.

**Comment by renpytom** [comment]
Do you plan to implement this?

**Comment by NintendoManiac64** [comment]
This was something that had been stewing in my head ever since it was announced back in June but sadly for a variety of reasons (including but not limited to PC hardware and software issues on my end that have become very apparent in the last couple weeks) it's not something I feel comfortable taking a stab at, at least not for quite a long long while.

So rather than leave the idea stewing in my head, I wanted to at least put it out there to see if others had any interest in the subject matter.

**Comment by LuqueDaniel** [comment]
To implement FSR it's required to use a compatible API (Vulkan and/or DirectX 11/12). Currently, it's not feasible to implement FSR in Ren'Py.

**Comment by renpytom** [comment]
I'm going to close this for now, as I don't really see this getting on the roadmap anytime soon, because of what @LuqueDaniel pointed out.

**Comment by mirh** [comment]
If you can run a shader on the gpu, then it will work. The thing has been [already](https://jntesteves.github.io/shadesofnoice/graphics/shaders/upscaling/2021/09/11/amd-fsr-demystified.html) shown to work in opengl. 
And even if you couldn't, you may still ship one of the aforementioned algorithms to run on the cpu.

**Comment by mirh** [comment]
https://github.com/yuzu-emu/yuzu/pull/9539
There was never any real hard lock.

**Comment by NintendoManiac64** [comment]
Huh, who new?  FSR on OpenGL *is* possible.

In that case, shouldn't this be reopened then?

**Comment by suphamster** [comment]
Here is comparison of default RenPy v8.1.0 upscaler vs [Magpie](https://github.com/Blinue/Magpie)
Magpie full screen mode with Anime4K Denoise UL scaler profile (from 720p to 2k):
![2714585_1686522953817](https://github.com/renpy/renpy/assets/17669188/13e8d464-9adb-4f75-8fbf-76f9a127095c)
Plain full screen (from 720p to 2k) mode (Alt+Enter):
![2714593_1686522964493](https://github.com/renpy/renpy/assets/17669188/ac53776c-9cb5-438c-a2a7-02611dbc480d)
It's easy to use on Windows 10/11 so I think this is the best free solution of proper upscaling for 2k/4k display users.

**Comment by NintendoManiac64** [comment]
It's not wise to rely on anime-optimized upscalers like Anime4K since you cannot guarantee that Ren'Py will be used with anime-style artwork, hence my suggestion of what is now generally known as FSR 1.0.

(also, 2K is "cinema 1080p" aka 2048x1080 - ignore all the people, online shops, and monitor makers that have been calling 1440p "2K" as it is wrong and has been factually been wrong from the beginning since anybody that understands what a resolution ending in "K" means also realizes why 1440p can't equal "2K" - consider that 4K equals ~4000px wide, typically 3840x2160 for 16:9 or 4096x2160 for 1.85 cinema. Half of 4 is 2, and half of 2160 is 1080 which, in 16:9 and 1.85 cinema equals 2048x1080 and 1920x1080. Furthermore, consider how 5K equals ~5000 pixels wide and, to date, basically all 5K displays are 5120px wide; half of 5 is 2.5, not 2, and half of 5120 is 2560 which is 2560x1440 aka 1440p for a 16:9 ratio - heck, pro cameras even shoot both 2K (1080p) and 2.5K (1440p) and label those resolutions as such because that's what they mean.)

**Comment by mirh** [comment]
I don't want to nag anybody, but at least the issue should be reopened
There's no stopper anywhere, the GLSL shaders were always there to begin with (despite what the readmes may suggest) and the ball is in this court

**Comment by NintendoManiac64** [comment]
I don't know if it's frowned upon, but I can't imagine it'd be the worst crime in the world to open another issue instead titled something like "Add yuzu's OpenGL implementation of FSR 1.0 for improved scaling".

Thing is, back when I originally created this, I didn't realize Ren'Py used OpenGL (in retrospect I should have realized that considering its native Linux support), and I have zero experience with OpenGL, so...

**Comment by renpytom** [comment]
Please don't open another issue unless you plan to be the one to implement it.

**Comment by NintendoManiac64** [comment]
> Please don't open another issue unless you plan to be the one to implement it.

I suppose that answers that question on whether it's frowned upon! Thanks for clearing that up. 👍

...but, um, shouldn't this still at least be reopened now that there is in fact an OpenGL implementation of FSR 1.0 since the reason this was closed was because, at the time, FSR required Vulkan or DirectX.

**Comment by mirh** [comment]
It is not yuzu implementation, and this was always possible to begin with (for the third time). 
If you rely on a fragment shader rather than compute ones like they did, you shouldn't even need to bump the system requirements.

**Comment by NintendoManiac64** [comment]
> If you rely on a fragment shader rather than compute ones like they did, you shouldn't even need to bump the system requirements.

And that's without even considering how GPU performance has improved in the 2.5 years since this issue was created; 1st generation Ryzen APUs as a whole were only around 3.5 years old at that point, but now they're 6 years old.

Heck, on the subject of emulators, shaders, and the performance of old GPUs, even my own Haswell Intel 4th gen integrated graphics are able to run Dolphin's infamous "exclusive ubershaders" (not hybrid ubershaders), albeit with only 1x native resolution, at around 35fps on the title screen／attract demo of FAST Racing League. And since Ivy Bridge Intel 3rd gen had quite similar integrated graphics, I wouldn't be surprised if that'd only be a little slower (it's Sandy Bridge Intel 2nd gen that had *much* worse intetegrated graphics).

While 30+ fps isn't the greating for a game like that whose native speed is 60, 2D visual novels would *easily* be perfectly fine at 30fps; heck they'd probably even be fine at 20fps.

**Comment by mirh** [comment]
You really should get those absurd digressions checked..

**Comment by mirh** [comment]
Other implementations (including gl < 4.0 compatibility and a RGB native version)
https://gist.github.com/agyild/82219c545228d70c5604f865ce0b0ce5
https://github.com/RPCS3/rpcs3/tree/master/rpcs3/Emu/RSX/Program/Upscalers/FSR1

**Comment by mirh** [comment]
That's [going](https://www.pcgamer.com/hardware/microsoft-details-how-its-automatic-super-resolution-upscaler-will-benefit-gamers-with-copilot-ai-pcs/) to use dedicated AI blocks for upscaling.. so perhaps the image may be better than just FSR (and being a post-process thing that adds extra latency isn't going to matter much for novels). 
Still, it's directx only (and I'm not even sure how it could work with a window that doesn't have a fixed fullscreen resolution)

Anyhow, there are at least 5 separate opengl implementations of FSR by now. Come on.

**Comment by NintendoManiac64** [comment]
> it seems that W11 and future OS [will try to embed this kind of technology natively](https://www.tomshardware.com/software/operating-systems/ai-super-resolution-scaling-for-all-apps-previewed-in-latest-windows-11-insider-build).

The problem with that it scales *everything* in the program window or the like, but Ren'Py for quite a while now has supported full resolution independence where it renders the text independently of the graphics, allowing for perfectly crisp text on a 4k display even if the art assets are a mere 640x480 or the like.

...hence the idea of integrated a higher quality general-purpose upscaler for the graphics, FSR 1.0 seeming to be one of the best candidates for that due to its performance (something that will become even more of a non-issue as time goes on).

**Comment by renpytom** [comment]
If someone wants to make a patch, I'll consider it.

**Comment by NintendoManiac64** [comment]
Oh by the way, I'm also a Linux user, so Windows upscaling doesn't help there. Yes Steam Deck has built-in FSR 1.0, but it provides the exact same limitations I'm describing.

> for this kind of upscaler you would want to run everything at native resolution (there is nothing preventing it, just disabling dpi scaling while the upscaler would be running).

That still doesn't solve the issue with the Windows scaler though because, I know since at least 2018, possibly earlier, text is rendered with resolution independence—that is, *every* resolution is native for Ren'py's text rendering. So if your visual novel has art assets that are only 800x600 (the VN standard for many many years), capping your program window to that resolution just to be able to take advantage of the Windows upscaler will quite negatively impact the clarity of your text.

**Comment by NintendoManiac64** [comment]
I think we're talking about two different things...

First off, the Windows scaler as well as the FSR built into the Steam Deck is purely a post-process spacial upscaling filter, most similar to the likes of what happens if you feed a 480p video signal to an HDTV—it will upscale that 480p signal to whatever the native resolution of your panel is such as 768p, 1080p, 2160p, etc.

This is purely what I'm objecting to—relying on the existence of external spacial upscalers.

Conversely, if the likes of FSR 1.0 were integrated into Ren'Py itself, then this is a complete non-issue because it would behave exactly like Ren'Py currently behaves with its present seemingly-bilinear upscaler. If we take the example of a modern VN running on Ren'Py 8 that has a native art asset resolution of 800x600 (they do in fact exist) then, when running such VN in fullscreen or a maximized window on basically any semi-modern display, lets say a big 'ol 3840x2400 display, the text will be natively rendered, *not upscaled*, at that 2400p resolution as if that were the VN's native resolution. However, all of the art assets will still be spatially upscaled using the likes of bilinear from 600p to 2400p which... won't look great to put it lightly.

**Comment by NintendoManiac64** [comment]
> However if the base resolution is too low, it's probably better to render the text at window resolution.

Let's just say there is a specific still-developed 800x600 VN using Ren'Py that I had in mind when I opened this issue initially. 

Also, is there a reason to *not* just always render the text at window resolution? My idea was more of a straight-up improvement/replacement of the current bilinear(?) upscaling, and not just *always* doing that would then over-complicate things with having multiple code-paths and such and/or possibly requiring input from the end-user in order to configure what method to be used or the like since you'd end up in a situation of trade-offs rather than a general all-around improvement.

> If someone wants to make a patch, I'll consider it.

Does it make sense for this issue to still be closed then? Maybe it's just me but a closed issue in my view implies the task has been completed in some fashion, whether that being something that indeed since been implemented or it's something that will *not* be considered for inclusion.

**Comment by renpytom** [comment]
I don't think the discussion on this thread is useful, so it should stay closed. 

If someone can do a proof of concept, then we can revisit it, but just chatting about what _can_ be done by someone else is rarely a useful thing.

**Comment by mirh** [comment]
The chat that didn't understand the difference between text and image upscaling was useless (btw magpie also has other AI-scalers that do a better job in brute forcing the entire window)
But the thing about finding an opengl shader seemed pretty poignant (especially considering the original reason this was closed). 

Then, putting aside that I'm not 100% sure if the entire thing could run with just GLES2 or it would require GLES3, it's not really a straightforward text effect. So.. first of all I don't think it could fit into `_shaders.rpym`, and secondly it's big enough that something like #3852 might be very preferably needed.

**Comment by AJolly** [comment]
FWIW I find myself unpacking renpy games, then rescaling all assets for 4k, (Waifu2x-Extension-Gui), which has a range of engines including RTX super resolution, then playing the game once thats done.  Takes a  bit of time, but its worth it (to me)

**Comment by suphamster** [comment]
> RTX super resolution

I don't think it worth efforts what you done since with Magpie build-in filters you can get similar results without converting images, I think it could only worth if you did it with Topaz or similar tool that has faces detection and other useful AI features.

**Comment by AJolly** [comment]
> I don't think it worth efforts what you done since with Magpie build-in filters you can get similar results without converting images, I think it could only worth if you did it with Topaz or similar tool that has faces detection and other useful AI features.

Thanks, I had not heard of magpie before! But yeah I've also played with Davinci Resolve
